### ISSUE TITLE
Memory leak for literal regexes since switch to enif_alloc/enif_free

### ISSUE THREAD

**Issue by tuncer** [tuncer/re2#17]
Compiled regexes correctly invoke the RE2 destructor on NIF resource cleanup, but the temporary RE2 objects constructed for literal regexes are not destructed, and we only free the associated erts memory block.

I'm working on a fix. Ideally, now that we require C++11 and RE2's constructors seem to be compatible, we could use unique_ptr/shared_ptr, which we weren't able to when the code was originally written. I might have to introduce a short-term fix first.

**Comment by tuncer** [comment]
@matehat @guitarmind @bernardd @junyer testing a potential short-term fix.

**Comment by tuncer** [comment]
Using the short-term fix branch, running `re2:match("foo", "o")` 3 million times, including the garbage of inefficiently using `lists:seq(1,N)` instead of an allocation-free bounded looping mechanism, went up quite high, but after all garbage had been collected, the VM settled at 43MB.

**Comment by tuncer** [comment]
@matehat @guitarmind, can you please test the `dtor` branch?

**Comment by guitarmind** [comment]
@tuncer  Thanks for the short-term fix!
Will pull `dtor` branch and test it locally, then give feedback to you as soon as possible!

**Comment by junyer** [comment]
The change in the `dtor` branch looks good to me and, of course, I'd be happy to discuss/review further changes.

**Comment by junyer** [comment]
A couple of thoughts so far:

1. We could replace the `autohandle<RE2>` with a `std::unique_ptr<RE2>` and a `RE2*`: the former owns the "temporary" `RE2` object (when one is required) and the latter points to the "temporary" `RE2` object or the previously compiled `RE2` object.

2. Does it matter that RE2 does not use `enif_alloc()` and `enif_free()`? If not, would it matter if the wrapper did not use them either? We could provide a deleter for the `std::unique_ptr<RE2>`, but it would be simpler not to have one.

**Comment by tuncer** [comment]
Added a commit that extracts the cleanup code into a shared function. It also adds a (commented out) deleter function object, but the conversion to unique_ptr isn't done yet. @guitarmind, please test the updated branch, and if it's all good, I'll cut a release this week. No need to delay it for the unique_ptr conversion.

**Comment by tuncer** [comment]
@junyer, I've had some trouble using it like `std::unique_ptr<re2::RE2, EnifDeleter>`, so I'll have to get back to it later.

(1): I need to see if I can use `unique_ptr`'s API (`release()`) similar to `autohandle<RE2>` to continue using a single variable (ptr), not selecting one or the other pointer.

(2): I'd prefer to keep using the VM's allocator, and using a custom allocator with `unique_ptr` seems straight forward (so far).

**Comment by guitarmind** [comment]
@tuncer @junyer @matehat 

FYI. I tried to run the short-term fix branch `dtor` but still observed memory leak at the same place. (`re2_match()+841  c_src/re2_nif.cpp:326`) [file link](https://github.com/tuncer/re2/blob/753d61d5219c6370730e2e3edf4e78730c9d10ac/c_src/re2_nif.cpp#L326), so the line number is different because it's a newer version)

My branch used for testing the `dtor` fix: [guitarmind/cqerl.git#markpeng/dtor](https://github.com/guitarmind/cqerl/tree/markpeng/dtor)

An example:
```
CallStack[8]: may-leak=11 (440 bytes)
    expired=15 (600 bytes), free_expired=4 (160 bytes)
    alloc=29834 (1193360 bytes), free=29823 (1192920 bytes)
    freed memory live time: min=0 max=14 average=0
    un-freed memory live time: max=258
    0x00007ff2b7d75750  libc-2.19.so  __libc_malloc()+0  /build/eglibc-3GlaMS/eglibc-2.19/malloc/hooks.c:77
    0x00007ff2b022adad  libstdc++.so  _Znwm()+29
    0x00007ff2b052054d  re2_nif.so  _ZN3re214SimplifyWalker7Concat2EPNS_6RegexpES2()+29
    0x00007ff2b0520abf  re2_nif.so  _ZN3re214SimplifyWalker14SimplifyRepeatEPNS_6R()+1311
    0x00007ff2b05217ef  re2_nif.so  _ZN3re214SimplifyWalker9PostVisitEPNS_6RegexpE()+463
    0x00007ff2b05245a7  re2_nif.so  _ZN3re26Regexp6WalkerIPS0_E12WalkInternalES2_S()+615
    0x00007ff2b0522d99  re2_nif.so  _ZN3re26Regexp8SimplifyEv()+137
    0x00007ff2b05032f7  re2_nif.so  _ZN3re28Compiler7CompileEPNS_6RegexpEbl()+71
    0x00007ff2b04e76f6  re2_nif.so  _ZN3re23RE24InitERKNS_11StringPieceERKNS0_7Opt()+502
    0x00007ff2b04e896d  re2_nif.so  _ZN3re23RE2C1ERKNS_11StringPieceERKNS0_7Option()+109
    0x00007ff2b04e5309  re2_nif.so  re2_match()+841  c_src/re2_nif.cpp:326
    0x0000000000436370  beam.smp  process_main()+16624
    0x00000000004b4fdf  beam.smp
    0x00000000005b92ab  beam.smp
    0x00007ff2b82c8182  libpthread-2.19.so  start_thread()+194  /build/eglibc-3GlaMS/eglibc-2.19/nptl/pthread_create.c:312
    0x00007ff2b7ded47d  libc-2.19.so  __clone()+109  ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

Here is the output of [memleax](https://github.com/WuBingzheng/memleax):
[mem_leak_test_dtor_20161005.txt](https://github.com/tuncer/re2/files/510229/mem_leak_test_dtor_20161005.txt)

**Comment by tuncer** [comment]
@guitarmind, I've added a simple stress tester to the branch.

```
$ rebar co
$ ./l.sh 1000000
28.93user 0.51system 0:30.36elapsed 97%CPU (0avgtext+0avgdata 24072maxresident)k
0inputs+0outputs (0major+5818minor)pagefaults 0swaps
$ /usr/bin/time erl -noinput -eval 'io:format("dtor~n"),init:stop().'
dtor
0.30user 0.03system 0:01.26elapsed 26%CPU (0avgtext+0avgdata 22160maxresident)k
0inputs+0outputs (0major+5581minor)pagefaults 0swaps
```

Above you can see the runtime difference between running 1 million iterations of `re2:replace` and `re:match` vs a simple print-and-exit of the VM.

I also ran 10 million iterations and also didn't get over the 24MB RAM used by the VM.

If I reintroduce the leak
```
diff --git a/c_src/re2_nif.cpp b/c_src/re2_nif.cpp
index 966b501..13d92b2 100644
--- a/c_src/re2_nif.cpp
+++ b/c_src/re2_nif.cpp
@@ -41,7 +41,7 @@ namespace {
     void cleanup_obj_ptr(T* ptr)
     {
         if (ptr != nullptr) {
-            ptr->~T();
+            //ptr->~T();
             enif_free(ptr);
             ptr = nullptr;
         }
```

an run another 1 million iterations, it's different:

```
$ ./l.sh 1000000
29.90user 2.93system 0:33.68elapsed 97%CPU (0avgtext+0avgdata 6154784maxresident)k
0inputs+0outputs (0major+1538568minor)pagefaults 0swaps
```

I'm not saying memleax is wrong, but when I accidentally forgot to call the RE2 destructor while converting to the VM allocator, it happened partly because Valgrind didn't report any issues in my incorrect code. As it turns out, I either didn't use Valgrind correctly, or it wasn't able to catch it back then. Basically, what I want to know is the memory footprint of your Erlang VM with latest `dtor` branch when running your code or when running `re2:l(10000000)` compared to `master` branch.

In [memleax#how-it-works](https://github.com/WuBingzheng/memleax#how-it-works) it says

> reports the blocks which live longer than 10 seconds

and

> NOTE: Since memleax does not run along with the whole life of target process, it assumes the long-lived memory blocks are memory leaks.

I'm not sure if that observation by memleax is assume to assume a leak. Maybe running it with a different `-e T` setting makes sense here.

**Comment by guitarmind** [comment]
@tuncer Good point about the memlex. I will test it again with larger `-e T` and also run your simple stress tester locally. But it might take some time to get back to you. Thanks again for the follow-up!

**Comment by tuncer** [comment]
@guitarmind, any idea when? I'd like to cut a release with the bugfix.

**Comment by guitarmind** [comment]
@tuncer The plan is to do it this weekend and get back to you, thanks.

**Comment by guitarmind** [comment]
@tuncer 
Just got some time to test it earlier. So the result looks quite good!
I tried to run both your stress tester and our service under the same environment, here are what I got:

**Stress Tester (without leak)**

1 million iterations
```
$ rebar co -DDEV
==> re2 (compile)
Compiled src/re2.erl
$ dev/l.sh 1000000
23.67user 0.49system 0:26.45elapsed 91%CPU (0avgtext+0avgdata 24180maxresident)k
808inputs+0outputs (75major+4273minor)pagefaults 0swaps
```

10 million iterations
```
$ dev/l.sh 10000000
224.23user 4.31system 3:56.62elapsed 96%CPU (0avgtext+0avgdata 25356maxresident)k
32inputs+0outputs (1major+4069minor)pagefaults 0swaps
```

**Stress Tester (with leak)**

10 million iterations
```
$ dev/l.sh 10000000
Command terminated by signal 9
6.54user 1.55system 0:11.32elapsed 71%CPU (0avgtext+0avgdata 1645492maxresident)k
401672inputs+0outputs (5127major+409655minor)pagefaults 0swaps
```
I can see it exploded at the speed of light...

**Integration Test in Our Service**

The memory usage of VM will settle down at around 170-180MB after running ~15 minutes.
I also opened memleax at the same time with `-e 600` to set memory free expire time to 10 minute, and this time the console output became empty (means no leak detection!)

In summary, the short-term fix has done a good job, thanks for the efforts.

**Comment by tuncer** [comment]
Thanks. Just released 1.6.0.
